language/tools/mirai-dataflow-analysis/analyses/ddlog.dl

```diff
@@ -270,20 +289,37 @@ TypedDataflowVia(node1, node2, node3, t) :- TypedDataflow(node1, node2, t), NotE
                                             TypedDataflow(node2, node3, t), NotEqual(node2, node3).
 
 
+// Safe dataflow
+// A dataflow from node1 to node2 (a checker) is considered safe if
+//   a. node1 is a checker
+//   b. node1 has a dataflow to node2
+//   c. node2 dominates node1 for t
+output relation SafeDataflow(node1: NodeID, node2: NodeID, t: TypeID)
```

Comment:
Discussed with Darion offline on why we need this `SafeDataflow(node1, node2, t)`, which is used in `TypedDataflowNotVia`.

Ideally we aim to capture that either
1. `node1` is on all data flows to `node2` w.r.t `t`, or
2. `node2` is on all data flows to `node1` w.r.t `t`.

But this is not what is exactly captured in this `SafeDataflow`. In particular:
- Case `b` is a match for rule 1, but
- Case `c` is a conservative instantiation of rule 2 (since dominates is only one way of data flow).

But since this is conservative, it is OK to land now. @darioncassel feel free to chime in and explain why we necessary need this relaxation with the examples you have.

In the discussion, another topic is brought up which is worth noting. The dominance relation is not corrected captured in cases of iterators, as shown in the following case:

```rust
fn entry(data) {
    check(data);
    data.map(|..| { <closure> });
}
```

The dominance relation from `check` to the `<closure>` function is not captured. And this might cause some false positives (sth is checked but marked us as not checked) in the `SafeDataflow` rule, and subsequently, `TypedDataflowNotVia`.

Reply:
I'll chime in to highlight the need for case `c` (which captures just dataflow via dominance) just rather than rule 2 (dataflow via dominance and calling). The issue is that we can have a scenario like this, where `node2` is a checking function (`checker`):
```
fn main() {
  let data: T = ...
  if (cond) {
    checker(data);
  } else {
    node1(data);
  }
}

function checker(data: T) {
  node1(data);
  // Perform the check on data
}
```
In this case, the call to the `checker` does not dominate `node1` and the `checker` calls `node1` with the data  _before_ the check is performed on the data. If our definition of SafeDataflow included dataflow via calling, we would claim that type `T` (the type of the data) is checked in this case when it is not.